### PR TITLE
Implementation for #126: supporting evaluation & forward feature selection for custom metrics (e.g. lift)

### LINKED ISSUE #126: Tuning classification model on different metrics (lift, recall, precision,...)
# Tuning classification model on different metrics (lift, recall, precision,...)

now a classification cobra model can only be tuned on AUC

## Task Description

This issue will...

### PR DESCRIPTION
**Pull request by sandervh14**
**Do not merge yet!** See below for explanation.

# Story Title

See [issue #126: Tuning classification model on different metrics (lift, recall, precision,...)](https://github.com/PythonPredictions/cobra/issues/126)

## Changes made

- Adapted forward feature selection class: 
  * it can now accept metric, metric_args and metric_kwargs and higher_is_better as argument. 
  * Extensive documentation of what to fill in in those parameters.
  * compute_model_performances: passing metric, metric_args and metric_kwargs to model evaluate()
  * _find_next_best_model(): idem + adapted logic for selecting new model as next best, according to higher_is_better logic.
- Adapted both LogisticRegressionModel and LinearRegressionModel to support a custom metric, metric_args and metric_kwargs in evaluate().

I tried out my changes in a notebook on a model, looks good. :-)
**But do not merge yet!**
I'm stuck on the 25 failing unit tests, mainly due to the cryptical error "mocker not found":

```
test setup failed
file C:\Users\sander.vandenhautte\PycharmProjects\cobra\tests\model_building\test_forward_selection.py, line 55
      @pytest.mark.parametrize("model_type", ["classification", "regression"])
      def test_compute_model_performances(self, mocker, model_type):
E       fixture 'mocker' not found
>       available fixtures: anyio_backend, anyio_backend_name, anyio_backend_options, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```

=> *Are* we using fixtures? With a find, I can't locate where `mocker` is instantiated.

Anyway, next to this question, of course fixing the unit tests remain, before we can merge this into develop.
But meanwhile Hendrik can use the issue branch, and I can discuss with you Sam how to fix the above unit test shizzle. :-)
I'll plan in that work for next week.

## How does the solution address the problem

This PR lets any data scientist the freedom to pick a different evaluation metric for forward feature selection or just for simple evaluation of a model.

## Linked issues

Resolves #126

### CODE REVIEW

**Review by sborms — COMMENTED** [review]
Well done, good additions! Looks almost finished, I just added a few documentation remarks, and some questions about the (code) distinction between y_pred, y_score, etc.

**Comment by sborms** [comment]
Wrt the mocking, you need to install pytest-mock (pip install pytest-mock).

File: cobra/model_building/forward_selection.py [hunk 1]
```diff
@@ -33,12 +39,61 @@ class ForwardFeatureSelection:
         Whether or not the model coefficients should all be positive (no sign flips).
     self._fitted_models : list
         List of fitted models.
+    metric : Callable (function), optional
+        Function that evaluates the model's performance, by calculating a
+        certain evaluation metric.
+        If the metric is not provided, the default metric AUC is used for
+        evaluating the model.
+        The metric functions from sklearn can be used, see
```
[the remark this answers is not in the document]
Comment (sandervh14) [reply cobra/model_building/forward_selection.py:47]: Haha woops copy pasta

File: cobra/model_building/forward_selection.py [hunk 2]
```diff
@@ -33,12 +39,61 @@ class ForwardFeatureSelection:
...
+    metric_kwargs : dict, optional
+        Keyword arguments (for example: normalize=True) to be passed to the
+        metric function when evaluating the model's performance.
+        Example metric function in which this is required (from
+        scikit-learn):
+        def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None)
+    higher_is_better : bool, optional
+        Whether the model is performing better if the chosen evaluation
+        metric results in a higher score (higher_is_better=True),
+        or worse (higher_is_better=False, meaning "lower is better").
```
[the remark this answers is not in the document]
Comment (sandervh14) [reply cobra/model_building/forward_selection.py:86]: Hi Sam, I've changed the longer argument documentations into referrals to other places instead of repeating the long texts. This one though I'd like to leave as is, to me someone calling this function and fetching the docstring for it, rather than digging through the code of the entire class, still needs to have some info to understand it. `higher_is_better: bool, optional See spot x.` would be too less info to me.

File: cobra/model_building/models.py [hunk 3]
```diff
@@ -390,34 +452,86 @@ def evaluate(self, X: pd.DataFrame, y: pd.Series,
...
-        metric: Callable (function), optional
-            Function that computes an evaluation metric to evaluate the model's
-            performances, instead of the default metric (RMSE).
-            The function should require y_true and y_pred arguments.
-            Metric functions from sklearn can be used, for example, see
+        metric : Callable (function), optional
+            Function that evaluates the model's performance, by calculating a
+            certain evaluation metric.
+            If the metric is not provided, the default metric AUC is used for
+            evaluating the model.
```
[the remark this answers is not in the document]
Comment (sandervh14) [reply cobra/model_building/models.py:459]: RMSE: OK.
Duplication: I decided to leave the documentation of "metric" in the models - unfortunately, yes, it's then still in 2 spots (both LinearRegressionModel.evaluate and LogisticRegressionModel.evaluate), but some of the explanation is dedicated to the type of the model - e.g. see below example) + I think the documentation of "metric" belongs more to the model (how you evaluate the model), rather than forward feature selection. So I removed it in forward feature selection, but left it in the 2 models, because of the deviations of the text specifically for that model:
```
Examples:
            - sklearn.metrics.r2_score(y_true, y_pred, *, sample_weight=None, multioutput='uniform_average')
            - overall_estimated_commission_earned(y_true, y_pred,
                                                  avg_prob_buy_if_err_lower_than_20K=0.25,
                                                  avg_prob_buy_if_err_higher_than_20K=0.05,
                                                  pct_commission_on_buy=0.05)
```

File: cobra/model_building/forward_selection.py [hunk 4]
```diff
@@ -33,12 +39,61 @@ class ForwardFeatureSelection:
...
+        through the metric_args and metric_kwargs parameters.
+        If you are unsure which arguments of your metric function are
+        args/kwargs, then run inspect.getfullargspec(your_metric_function).
+    metric_args : dict, optional
+        Arguments (for example: lift_at=0.05) to be passed to the metric
+        function when evaluating the model's performance.
+        Example metric function in which this is required:
+        ClassificationEvaluator._compute_lift(y_true=y_true,
+                                              y_score=y_pred,
+                                              lift_at=0.05)
```
[the remark this answers is not in the document]
Comment (sandervh14) [reply cobra/model_building/forward_selection.py:76]: Well noticed. :-) Had to be y_score=y_score :-)

File: cobra/model_building/models.py [hunk 5]
```diff
@@ -163,39 +165,97 @@ def evaluate(self, X: pd.DataFrame, y: pd.Series,
...
+            # now, with any evaluation metric possibly being used, y_pred may
+            # be required instead of y_score, which requires determining the
+            # optimal threshold first and then calculating y_pred.
+            fpr, tpr, thresholds = roc_curve(y_true=y, y_score=y_score)
+            cutoff = ClassificationEvaluator._compute_optimal_cutoff(fpr, tpr,
+                                                                     thresholds)
+            y_pred = np.array([0 if score <= cutoff
+                               else 1
+                               for score in y_score])
+
```
[the remark this answers is not in the document]
Comment (sandervh14) [reply cobra/model_building/models.py:229]: > If I understand well, if the metric function doesn't need y_pred, these calculations are redundant? Maybe try to catch the metric functions with an y_pred argument before running this.

Good thinking! I've added that.

> I also don't see y_proba documented, is this needed?

I noticed two of [sklearn's metric functions](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) use that argument name (actually y_prob instead of y_proba, modified that) rather than y_score and added it for the sake of completeness. I've added this explanation in the docstring: "The variables that we provide and your function can possibly take in are y_true, y_pred, y_score and y_prob."

File: cobra/model_building/models.py [hunk 6]
```diff
@@ -390,34 +452,86 @@ def evaluate(self, X: pd.DataFrame, y: pd.Series,
...
+        else:
+            # A custom evaluation metric was chosen. With the default
+            # metric RMSE, the performance could be scored over all possible
+            # based on y_score;
+            # now, with any evaluation metric possibly being used, y_pred may
+            # be required instead of y_score, so we'll first calculate y_pred.
+            # Compute the model performance with the chosen metric function,
+            # pass all arguments this function could potentially need,
+            # including optional keyword arguments that were passed when
+            # initializing this model.
```
[the remark this answers is not in the document]
Comment (sandervh14) [reply cobra/model_building/models.py:511]: Hm, you're right. I copied this over from the LogisticRegressionModel (classification) without thinking carefully, for regression indeed y_true and y_pred are enough.

**Comment by sandervh14** [comment]
> Wrt the mocking, you need to install pytest-mock (pip install pytest-mock).

Thanks, I've added that info to the [Contributing guidelins & workflow](https://github.com/PythonPredictions/cobra/wiki/Contributing-guidelines-&-workflows) for in case other will have the same question as I did.

**Comment by sandervh14** [comment]
Hey Sam, just FYI: I'm committing my code changes I did based on your review, but then some stuff came up in between because I'm going to start at a new client, I couldn't finish the unit tests, will do that after the long weekend.

File: cobra/model_building/models.py [hunk 7]
```diff
@@ -368,102 +415,201 @@ def score_model(self, X: pd.DataFrame) -> np.ndarray:
...
         return True
+
+    def _evaluate_with_default_metric(self, y_true, y_score):
+        """We use RMSE as default evaluation metric."""
+        return sqrt(
+            mean_squared_error(y_true=y_true,
+                               # LinearRegressionModel actually returns y_pred
+                               # inside y_score:
+                               y_pred=y_score)
+        )
```
Comment (ZlaTanskY) [inline cobra/model_building/models.py:607]:
```suggestion
        return  mean_squared_error(y_true=y_true,
                               # LinearRegressionModel actually returns y_pred
                               # inside y_score:
                               y_pred=y_score,
                               squared=False)
        )
```
